core/trino-main/src/main/java/io/trino/execution/AddColumnTask.java

```diff
@@ -92,49 +99,121 @@ public ListenableFuture<Void> execute(
+                    true);
+
+            ColumnMetadata column = ColumnMetadata.builder()
+                    .setName(columnName.getValue())
+                    .setType(type)
+                    .setNullable(element.isNullable())
+                    .setComment(element.getComment())
+                    .setProperties(columnProperties)
+                    .build();
+
+            plannerContext.getMetadata().addColumn(session, tableHandle, qualifiedTableName.asCatalogSchemaTableName(), column);
         }
-        if (!element.isNullable() && !plannerContext.getMetadata().getConnectorCapabilities(session, catalogHandle).contains(NOT_NULL_COLUMN_CONSTRAINT)) {
-            throw semanticException(NOT_SUPPORTED, element, "Catalog '%s' does not support NOT NULL for column '%s'", catalogHandle, element.getName());
+        else {
+            accessControl.checkCanAlterColumn(session.toSecurityContext(), qualifiedTableName);
+
+            if (!element.isNullable()) {
+                throw semanticException(NOT_SUPPORTED, statement, "Adding fields with NOT NULL constraint is unsupported");
+            }
+            if (!element.getProperties().isEmpty()) {
+                throw semanticException(NOT_SUPPORTED, statement, "Adding fields with column properties is unsupported");
+            }
+            if (element.getComment().isPresent()) {
+                throw semanticException(NOT_SUPPORTED, statement, "Adding fields with COMMENT is unsupported");
+            }
+            if (!columnHandles.containsKey(columnName.getValue().toLowerCase(ENGLISH))) {
+                throw semanticException(COLUMN_NOT_FOUND, statement, "Column '%s' does not exist", columnName);
+            }
+
+            List<String> parentPath = statement.getColumn().getField().getOriginalParts().subList(0, statement.getColumn().getField().getOriginalParts().size() - 1).stream()
+                    .map(Identifier::getValue)
+                    .collect(toImmutableList());
+            List<String> fieldPath = statement.getColumn().getField().getOriginalParts().subList(1, statement.getColumn().getField().getOriginalParts().size()).stream()
+                    .map(Identifier::getValue)
+                    .collect(toImmutableList());
+
+            ColumnMetadata columnMetadata = plannerContext.getMetadata().getColumnMetadata(session, tableHandle, columnHandles.get(columnName.getValue().toLowerCase(ENGLISH)));
+            Type currentType = columnMetadata.getType();
+            int fieldDepth = 0;
+            for (String fieldName : fieldPath) {
+                fieldDepth++;
+                if (currentType instanceof RowType rowType) {
+                    List<RowType.Field> candidates = rowType.getFields().stream()
+                            // case-sensitive match
+                            .filter(rowField -> rowField.getName().isPresent() && rowField.getName().get().equals(fieldName))
```

Comment:
We should not rely on `Identifier.getName()`. This method breaks the SQL identifier semantics as it ignores quotations.

Reply:
The `rowField` comes from the table definition, not comes from user input. I don't think it matters.